pkg/store/bucket.go

```diff
@@ -2325,112 +2321,228 @@ func ExpandPostingsWithContext(ctx context.Context, p index.Postings) (res []sto
+	if pg.name != other.name {
+		return nil
+	}
+	// Both add all, merge remove keys.
+	if pg.addAll && other.addAll {
+		// If nothing to merge, end early.
+		if len(pg.removeKeys) == 0 && len(other.removeKeys) == 0 {
+			return &pg
+		}
+		if len(other.removeKeys) == 0 {
+			return &pg
+		}
+		if len(pg.removeKeys) == 0 {
+			pg.removeKeys = other.removeKeys
+			return &pg
+		}
+		set := make(map[string]struct{}, len(pg.removeKeys)+len(other.removeKeys))
+		for _, key := range pg.removeKeys {
+			set[key] = struct{}{}
+		}
+		for _, key := range other.removeKeys {
+			set[key] = struct{}{}
+		}
+		pg.removeKeys = make([]string, len(set))
+		i := 0
+		for key := range set {
+			pg.removeKeys[i] = key
+			i++
+		}
+	} else if pg.addAll || other.addAll {
+		// Subtract the remove keys.
+		toRemove := other
+		toAdd := &pg
+		if pg.addAll {
+			toRemove = &pg
+			toAdd = other
+		}
+		set := make(map[string]struct{}, len(toRemove.removeKeys))
+		for _, key := range toRemove.removeKeys {
+			set[key] = struct{}{}
+		}
+		i := 0
+		for _, key := range toAdd.addKeys {
+			if _, ok := set[key]; !ok {
+				toAdd.addKeys[i] = key
+				i++
+			}
+		}
+		toAdd.addKeys = toAdd.addKeys[:i]
+		if pg.addAll {
+			pg.removeKeys = nil
+			pg.addKeys = other.addKeys
+			pg.addAll = false
+		}
+	} else {
+		set := make(map[string]struct{}, len(pg.addKeys))
+		for _, key := range pg.addKeys {
+			set[key] = struct{}{}
+		}
+		mergedKeys := make([]string, 0)
+		for _, key := range other.addKeys {
+			if _, ok := set[key]; ok {
+				mergedKeys = append(mergedKeys, key)
+			}
+		}
+		pg.addKeys = mergedKeys
+	}
+	return &pg
+}
+
+func checkNilPosting(name, value string, p index.Postings) index.Postings {
 	if p == nil {
 		// This should not happen. Debug for https://github.com/thanos-io/thanos/issues/874.
-		return index.ErrPostings(errors.Errorf("postings is nil for %s. It was never fetched.", l))
+		return index.ErrPostings(errors.Errorf("postings is nil for {%s=%s}. It was never fetched.", name, value))
 	}
 	return p
 }
 
+func matchersToPostingGroups(ctx context.Context, lvalsFn func(name string) ([]string, error), ms []*labels.Matcher) ([]*postingGroup, error) {
+	matchersMap := make(map[string][]*labels.Matcher)
+	for _, m := range ms {
+		matchersMap[m.Name] = append(matchersMap[m.Name], m)
+	}
+
+	pgs := make([]*postingGroup, 0)
+	// NOTE: Derived from tsdb.PostingsForMatchers.
+	for _, values := range matchersMap {
+		var (
+			mergedPG     *postingGroup
+			pg           *postingGroup
+			vals         []string
+			err          error
+			valuesCached bool
+		)
+		lvalsFunc := lvalsFn
+		// Merge PostingGroups with the same matcher into 1 to
+		//  avoid fetching duplicate postings.
+		for _, val := range values {
+			pg, vals, err = toPostingGroup(ctx, lvalsFunc, val)
+			if err != nil {
+				return nil, errors.Wrap(err, "toPostingGroup")
+			}
+			// Cache label values because label name is the same.
+			if !valuesCached && vals != nil {
```

Comment:
do we need the `valuesCached` var? 
Can't we always override the `lvalsFunc`?

Reply:
Yeah we can do this. I am just don't want to create a new func every time...
Idk how much it costs though. If it is cheap or similar to the `valuesCached` I added then I can override every time.

Reply:
another possibility is  `i == 0`?
This is a nit though.